pydomcfg/accessor.py

```diff
@@ -43,12 +56,97 @@ def jpk(self) -> int:
+            raise ValueError("`jpk` MUST be >= 0 (use 0 to unset jpk)")
         self._jpk = value
 
+    @property
+    def nml_ref_path(self) -> str:
+        """
+        Path to reference namelist.
+
+        Returns
+        -------
+        str
+        """
+        return self._nml_ref_path
+
+    @nml_ref_path.setter
+    def nml_ref_path(self, value: str):
+        self._nml_ref_path = value
+
+    # domzgr methods
     @_jpk_check
     def zco(self, *args: Any, **kwargs: Any) -> Dataset:
         return Zco(self._obj, self._jpk)(*args, **kwargs)
 
     zco.__doc__ = Zco.__call__.__doc__
+
+    # Emulate NEMO DOMAINcfg tools
+    def from_namelist(self, nml_cfg_path_or_io: Union[str, Path, IO[str]]):
+        """
+        TODO
+        """
+
+        # Parse and return a ChainMap
+        nml = self._namelist_parser(nml_cfg_path_or_io)
+
+        # Pick zgr class
+        if nml["ln_zco"]:
+            zgr_class = Zco
+        else:
+            raise NotImplementedError("Only `Zco` has been implemented so far.")
+
+        # Inspect the __call__ signatures and get the approriate variables
+        # from the namelists.
+        parameters = inspect.signature(zgr_class.__call__).parameters
+        kwargs = {key: nml[key] for key in parameters if key != "self"}
+        return Zco(self._obj, nml["jpkdta"])(**kwargs)
+
+    def _namelist_parser(
+        self, nml_cfg_path_or_io: Union[str, Path, IO[str]]
+    ) -> ChainMap:
+
+        if not HAS_F90NML:
+            raise ImportError(
+                "`f90nml` MUST be installed" " to use `obj.domcfg.from_namelist()`"
+            )
+
+        if not self.nml_ref_path:
+            raise ValueError(
+                "Set `nml_ref_path` before calling `obj.domcfg.from_namelist()`"
+                " For example: obj.domcfg.nml_ref_path = 'path/to/nml_ref'"
+            )
+
+        if self.jpk:
+            warnings.warn(
+                "`obj.domcfg.jpk` is ignored." " `jpk` is inferred from the namelists."
+            )
+
+        # Read namelists: cfg overrides ref
+        # TODO:
+        #   If we specify the third argument, the resulting namelist is written
+        #   The advantage is that it is formatted as the reference namelist
+        #   We could do it to a tmp file, then add the resulting string to the
+        #   global attributes of the final dataset.
+        nml_cfg = f90nml.read(nml_cfg_path_or_io)
+        nml = f90nml.patch(self.nml_ref_path, nml_cfg)
+
+        # Chain all nam blocks
+        # TODO:
+        #   Using ChainMap we disregard nam blocks, so assume there aren't
+        #   variables with the same name in different nam blocks. Is it OK?
+        #   If NO, maybe it's OK if we just select relevant blocks?
+        #   ChainMap would make life much easier...
+        #   We just look at the zgr call signatures, and we extract the variables
+        #   with the same name from any block.
+        chained = ChainMap(*nml.todict().values())
```

Comment:
@jdha @oceandie I didn't check if variables with the same name are used in multiple blocks... Thoughts?

Reply:
Sorry - I've been reviewing/editing a few papers this week so haven't had time to look at the namelists. For the domcfg namelist I think you should be ok as I don't think there are any, however in the main NEMO namelists there are duplications like `cn_dir` ... in meld_nml I've just kept the namblocks as the first dictionary key... I'll have a look at the start of next week - but for the purposes of testing I don't think you should worry too much. We can adjust later if something more formal is written for handling namelists.

Reply:
Sounds like we can just use ChainMap and forget about blocks. We can always change if we'll need to, but I think it's unlikely.

> I've just kept the namblocks as the first dictionary key

Yes, that's what's done here as well before chaining. `f90nml.read()` basically creates a dictionary mapping namblocks to variable dictionaries.

>  For the purposes of testing I don't think you should worry too much

This is not just testing, is the API to use namelists rather than python classes and arguments.

Reply:
Sorry - that was clumsy use of language. It's not a big step to adjust the handling of the namelists if the use of ChainMap is deemed inappropriate at a later stage. [in my head this period of laying down code is _testing_] ... we can discuss tomorrow.